core/src/util/shaderProgram.cpp

```diff
@@ -278,6 +279,24 @@ void ShaderProgram::invalidateAllPrograms() {
     
 }
 
+void ShaderProgram::setUniform(const std::string& _name, Texture& _texture) {
+    use();
+    GLint location = getUniformLocation(_name);
+    GLuint unit;
+    
+    if (m_textureSlots.find(_name) == m_textureSlots.end()) {
+        unit = m_freeTextureUnit;
+        m_textureSlots[_name] = { _texture.getGlHandle(), m_freeTextureUnit++ };
+    } else {
+        unit = m_textureSlots[_name].second;
+    }
+    
+    _texture.update(unit);
+    _texture.bind(unit);
+    
+    glUniform1i(location, unit);
+}
+
```

Comment:
I see, this would probably need some more thoughts to be solved. What I would do to have this fixed is solving the dependencies shader/texture by having a "uniform request" instead of an actual uniform attachment, and before each draw we solve the best unit/textureId combination for each texture for the shaders that needed it.